backend/src/composite/mod.rs

```diff
@@ -127,33 +177,47 @@ impl<'a, F: FieldElement> Backend<'a, F> for CompositeBackend<'a, F> {
                             .map(|(symbol, _)| symbol),
                     );
 
-                    backend
-                        .prove(&witness, None, witgen_callback)
-                        .map(|proof| (machine.clone(), proof))
+                    backend.prove(&witness, None, witgen_callback)
                 })
                 .collect::<Result<_, _>>()?,
         };
-        Ok(serde_json::to_vec(&proof).unwrap())
+        Ok(bincode::serialize(&proof).unwrap())
     }
 
     fn verify(&self, proof: &[u8], instances: &[Vec<F>]) -> Result<(), Error> {
-        let proof: CompositeProof = serde_json::from_slice(proof).unwrap();
-        for (machine, machine_proof) in proof.proofs {
+        let proof: CompositeProof = bincode::deserialize(proof).unwrap();
+        for (machine, machine_proof) in self.machine_names.iter().zip(proof.proofs) {
             let machine_data = self
                 .machine_data
-                .get(&machine)
+                .get(machine)
                 .ok_or_else(|| Error::BackendError(format!("Unknown machine: {machine}")))?;
             machine_data.backend.verify(&machine_proof, instances)?;
         }
         Ok(())
     }
 
-    fn export_setup(&self, _output: &mut dyn io::Write) -> Result<(), Error> {
-        unimplemented!()
+    fn export_setup(&self, output: &mut dyn io::Write) -> Result<(), Error> {
+        // All backend are the same, just pick the first
+        self.machine_data
+            .values()
+            .next()
+            .unwrap()
+            .backend
+            .export_setup(output)
```

Comment:
But couldn't it be the case that different proofs have different setups?

Reply:
Not the way the composite backend is implemented now: The `CompositeBackendFactory` instantiates the *same* backend for each machine. In the future, we could relax that, and then you'd be right!